### PR TITLE
Don`t apply checksum wrapper to mixed r/w streams

### PR DESCRIPTION
**Pull request by IljaN**
## Description
Don`t apply checksum wrapper on mixed read/write streams because this could corrupt the checksum context if the stream is for example seeked or rewinded.

## Motivation and Context
Might be the potential source of checksum issues we are having. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by IljaN** [comment]
Thinking this further, if a read and write occurs on the same stream we might just want to reset the checksum and don`t compute one at all because incorrect checksum will occur in this scenario anyways...

**Comment by guruz** [comment]
I'm not a PHP guy, but maybe it's also feasible to check the ```$mode``` in ```stream_open``` and only allow reading/writing based on the mode.
(And assume it's not ```rw```or whatever)?

**Comment by IljaN** [comment]
@guruz Yeah good plan need to search trough the codebase where we open rw streams

**Comment by PVince81** [comment]
How would it be possible to read from a stream being written already ?

Usually that's only possible if fopen is called with a special mode.
Normally all known code paths use "r" or "w" but if there are that use something else we should simply disable computing checksums, maybe even refrain from applying the checksum wrapper.

**Comment by IljaN** [comment]
I think this was guruz`s idea... also another potential pitfall which comes to mind might be rewind(), e.g read, rewind, read again...

File: lib/private/Files/Stream/Checksum.php [hunk 1]
```diff
@@ -136,7 +136,7 @@ private function updateHashingContexts($data) {
 	private function stripPartialFileExtension($path) {
 		$extension = \pathinfo($path, PATHINFO_EXTENSION);
 
-		if ($extension === 'part') {
+		if ( $extension === 'part') {
```
Comment (patrickjahns) [inline lib/private/Files/Stream/Checksum.php]: why?

File: tests/lib/Files/Storage/Wrapper/ChecksumTest.php [hunk 2]
```diff
@@ -46,40 +48,43 @@ public function setUp() {
...
-		$metaData = $this->instance->getMetaData('/foo.txt');
+		$metaData = $this->instance->getMetaData($this->testFileName);
 
 		$this->assertArrayHasKey('checksum', $metaData);
 		$this->assertEquals(self::EXPECTED_CHECKSUMS, $metaData['checksum']);
 	}
 
 	public function testReadFromFileHandleOnNewFileCalculatesChecksum() {
+		$this->sourceStorage->file_put_contents($this->testFileName, self::TEST_DATA);
+        // Make file available in filecache (without checksum)
```
Comment (patrickjahns) [inline tests/lib/Files/Storage/Wrapper/ChecksumTest.php]: indentation

**Comment by IljaN** [comment]
@ownclouders rebase

**Review by PVince81 — APPROVED** [review]
Code looks good :+1:

**Comment by PVince81** [comment]
please resolve the conflict

**Comment by IljaN** [comment]
Conflict resolved

**Comment by phil-davis** [comment]
Backport ``stable10`` #31619

**Pull request merged**